### ISSUE TITLE
[BUG] New k-means is extremely slow to build

### ISSUE THREAD

**Issue by cjnolet** [rapidsai/raft#762]
I've had to comment out the building of the tests for the new k-means implementation in the meantime because it is dreadfully slow to build. I'm not sure if it's because it's having to build several different specializations of the distance / fused l2_nn primitives or because of the heavy use of mdarray/mdspan, but we should figure out the culprit and get it building faster.

**Comment by cjnolet** [comment]
Referencing PR where this was commented out: https://github.com/rapidsai/raft/pull/761/files

**Comment by rg20** [comment]
Same issue when trying this in cuopt as well.

**Comment by cjnolet** [comment]
Linking to #354

**Comment by cjnolet** [comment]
I've verified that this issue is almost entirely from the call to `raft::distance::pairwise_distance`. Commenting it out in the new kmeans (see below) allows the tests to build in about 1-2min. Uncommenting it causes the corresponding tests to take in upwards of 10-15min on my workstation:

```c++
void pairwise_distance_kmeans(const raft::handle_t& handle,
                              const raft::device_matrix_view<const DataT, IndexT> X,
                              const raft::device_matrix_view<const DataT, IndexT> centroids,
                              const raft::device_matrix_view<DataT, IndexT> pairwiseDistance,
                              rmm::device_uvector<char>& workspace,
                              raft::distance::DistanceType metric)
{
//  auto n_samples  = X.extent(0);
//  auto n_features = X.extent(1);
//  auto n_clusters = centroids.extent(0);
//
//  ASSERT(X.extent(1) == centroids.extent(1),
//         "# features in dataset and centroids are different (must be same)");
//
//
//  raft::distance::pairwise_distance(handle,
//                                    X.data_handle(),
//                                    centroids.data_handle(),
//                                    pairwiseDistance.data_handle(),
//                                    n_samples,
//                                    n_clusters,
//                                    n_features,
////                                    workspace,
//                                    metric);
}
```

**Comment by cjnolet** [comment]
Also linking in #539

**Comment by lowener** [comment]
I also noticed that the `pairwise_distance` function was taking most of the time.
By doing a few tests, I saw that if we restrict `pairwise_distance` to the following metrics: `L2Expanded L2SqrtExpanded CosineExpanded L1 L2Unexpanded L2SqrtUnexpanded Linf` the compilation time is around 2 minutes. And if you add `HellingerExpanded LpUnexpanded Canberra` to the metrics list, the compilation time jumps to around 10 minutes.

The `LpUnexpanded` metric takes the most time with ~6 minutes of compilation on my machine.

A solution might be to create a `pairwise_distance_common` function that is only using basic metrics, because I don't think that all the metrics that we have are interesting to use with kmeans (for example Jaccard, Dice or KLDivergence).
And if at some point there is a need for an additional metric for kmeans we can also create a kmeans with the metric as a template, to directly call the interesting metric.
